// nse-bse-api/dist-cjs/nse/api/historical-api.js
"use strict";
/**
 * Historical data API methods for NSE
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.HistoricalApi = void 0;
const index_js_1 = require("../constants/index.js");
const date_formatter_js_1 = require("../utils/date-formatter.js");
class HistoricalApi {
    httpClient;
    constructor(httpClient) {
        this.httpClient = httpClient;
    }
    /**
     * Fetch equity historical data
     */
    async fetchEquityHistoricalData(params) {
        const { symbol, from_date, to_date, series = ["EQ"] } = params;
        const to = to_date ?? new Date();
        const from = from_date ?? new Date(to.getTime() - 30 * 86400000);
        (0, date_formatter_js_1.validateDateRange)(from, to);
        const chunks = (0, date_formatter_js_1.splitDateRange)(from, to, 100);
        const results = [];
        // Fetch data for each series type
        for (const seriesType of series) {
            for (const [cFrom, cTo] of chunks) {
                const response = await this.httpClient.request(`${index_js_1.BASE_URL}/NextApi/apiClient/GetQuoteApi`, {
                    functionName: "getHistoricalTradeData",
                    symbol,
                    series: seriesType,
                    fromDate: (0, date_formatter_js_1.formatDateDMY)(cFrom),
                    toDate: (0, date_formatter_js_1.formatDateDMY)(cTo),
                });
                // Handle different response structures
                const data = response?.data ?? response ?? [];
                const arr = Array.isArray(data) ? data.slice().reverse() : [];
                results.push(...arr);
            }
        }
        return results;
    }
    /**
     * Fetch VIX historical data
     */
    async fetchHistoricalVixData(params = {}) {
        const to = params.to_date ?? new Date();
        const from = params.from_date ?? new Date(to.getTime() - 30 * 86400000);
        (0, date_formatter_js_1.validateDateRange)(from, to);
        const chunks = (0, date_formatter_js_1.splitDateRange)(from, to, 365);
        const data = [];
        for (const [cFrom, cTo] of chunks) {
            const result = await this.httpClient.request(`${index_js_1.BASE_URL}/historicalOR/vixhistory`, {
                from: (0, date_formatter_js_1.formatDateDMY)(cFrom),
                to: (0, date_formatter_js_1.formatDateDMY)(cTo),
            });
            data.push(...result.data);
        }
        return data;
    }
    /**
     * Fetch F&O historical data
     */
    async fetchHistoricalFnoData(params) {
        const { symbol, instrument = "FUTIDX", from_date, to_date, expiry, option_type, strike_price } = params;
        const instrumentType = instrument.toUpperCase();
        const to = to_date ?? new Date();
        const from = from_date ?? new Date(to.getTime() - 30 * 86400000);
        (0, date_formatter_js_1.validateDateRange)(from, to);
        const query = {
            functionName: "getDerivativesHistoricalData",
            symbol: symbol.toUpperCase(),
            instrumentType,
        };
        if (expiry) {
            query.expiryDate = (0, date_formatter_js_1.formatDateExpiry)(expiry);
            query.year = expiry.getFullYear().toString();
        }
        if (instrumentType === "OPTIDX" || instrumentType === "OPTSTK") {
            if (!option_type) {
                throw new Error("`option_type` param is required for Stock or Index options.");
            }
            query.optionType = option_type;
            if (strike_price != null)
                query.strikePrice = strike_price;
        }
        const chunks = (0, date_formatter_js_1.splitDateRange)(from, to, 365);
        const data = [];
        for (const [cFrom, cTo] of chunks) {
            query.fromDate = (0, date_formatter_js_1.formatDateDMY)(cFrom);
            query.toDate = (0, date_formatter_js_1.formatDateDMY)(cTo);
            const response = await this.httpClient.request(`${index_js_1.BASE_URL}/NextApi/apiClient/GetQuoteApi`, query);
            const result = response?.data ?? response ?? [];
            if (Array.isArray(result)) {
                data.push(...result);
            }
        }
        return data;
    }
    /**
     * Fetch F&O underlying information
     */
    async fetchFnoUnderlying() {
        const data = await this.httpClient.request(`${index_js_1.BASE_URL}/underlying-information`);
        return data.data;
    }
    /**
     * Fetch index names
     */
    async fetchIndexNames() {
        return await this.httpClient.request(`${index_js_1.BASE_URL}/index-names`);
    }
}
exports.HistoricalApi = HistoricalApi;
